### ISSUE TITLE
Ability to emit logs to the framework of user's choice directly

### ISSUE THREAD

**Issue by petemounce** [aws/aws-sdk-net#311]
It would be really useful if the SDK could surface the strongly-typed `RequestMetrics` instance to the logging subsystem (and, actually, any other telemetry) - or if there were a way to plug in an adapter to the SDK's logging system. I happen to use NLog for logging, and I emit logs in a JSON format optimised for simplicity of harvest (via nxlog) to ingest into Elasticsearch by way of LogStash.

In my use-case, this means the application logs one JSON event per line. The logging library adds JSON attributes like timestamp, process id, thread id, and so forth, via configuration.

I think because of the AWS SDK's assumption that it's in control of formatting the string that ends up in log files as the `message` token (eg `{timestamp}|{processid}|{message}` as a contrived human-readable example), that means users that want control over the format of their log lines (as opposed to the message _within_ their log lines) have to eat another deserialisation/enrich/serialise cycle, or do something a little intrusive to the SDK.

Since the SDK is also responsible for retrying requests internally (I don't think it's possible to control that behaviour), it also emits logs (via the standard formatter) that are still hard to parse because in the event of retries, things turn into arrays (rather than are emitted as arrays with single items).

Would it be possible to take the idea of a custom log message formatter (introduced in #157), and re-apply it so that the log-framework itself can be pluggable? That would enable both the specific thing I'd like to do, and also other users to direct logs to the framework of their choice.

I was thinking about something like
```xml
<aws ...>
  <logging logTo="custom,KeepTheArrayButIWouldntNeedIt" customLogAdapter="type name, assembly"/>
</aws>
```

and the adapter's signature being something like:
```csharp
public interface ILogEvents {
  void Log(LogEvent @event); // LogEvent is assumed to have log-level as a property
}

public abstract class AwsSdkLogger : ILogEvents {
  public virtual void Log(LogEvent @event) { // allow override if the adapter wants to completely differ, and write structured logs
    var line = MapEventToLogLine(@event);
    WriteLogLine(line);
  }

  protected virtual string MapEventToLogLine(LogEvent @event) {
    //return the string as the current implementation does
    //this is where a user could override if they wanted a different string in their message token in their log lines - eg, remove the newlines that the SDK emits
  }

  protected virtual void WriteLogLine(string line) {
    // write the log to whatever sink
    // this is where a user could override to plug their logger in directly, rather than going via System.Diagnostics
  }
}
```

At that point, the logging concerns of the SDK are logically separate and really easy to customise to fit gracefully into the host application's existing setup.

For example, NLog sketch
```csharp
public class NLogAWSSDKLogAdapter : AwsSdkLogger {
  public class NLogAWSSDKLogAdapter() {
    _logger = LogManager.GetLogger("Amazon");
  }
  public override void Log(LogEvent @event) {
    var nlEvent = MapToNLog(@event);
    _logger.Log(nlEvent);
  }
  public NLog.LogEvent(LogEvent @event) {
    var e = new NLog.LogEvent();
    e.Properties["Attempts"] = PullOutRequestAttemptsData(@event);
    e.Properties["aws_ServiceName"] = @event.ServiceName;
    /// etc, on the basis that NLog itself will now be in charge of rendering a string for the event with its own rich layoutrenderers and layouts per target, so different log sinks can receive different messages as appropriate (eg, hipchat gets a short note and a link to a kibana dashboard, whereas logstash gets a fully structured JSON blob that it will pull apart into Elasticsearch)
    return e;
  }
}
```

It's crucial, really, for the value of this feature, that the change make available the log-event instances (IRequestMetrics and any others) to the adapter, and delegate serialisation / formatting entirely. If the consumer needs to deal with ETL, we won't have improved the current implementation.

NLog and other frameworks also have memory log-sinks, which is invaluable for testing - one can both read logs within test automation, but also assert against their content without needing to wrap standard out (for example).

**Comment by PavelSafronov** [comment]
What you are looking to do is actually possible right now, though it is not quite as straightforward as your suggestion. When the SDK performs logging, it passes all the instances that you are looking for (IRequestMetrics, etc.) to the logger, so the formatting is done only by the logger (if at all, in the case of in-memory loggers). For instance, take a look at the various [TraceData calls that InternalSystemDiagnosticsLogger makes](https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/Amazon.Runtime/Internal/Util/_bcl/Logger.Diagnostic.cs#L48-L66) and note that we are passing a LogMessage to the logger. It's possible to take advantage of this fact, as outlined in the two code samples below.

First we have TestTraceListener, a TraceListener which attempts to extract IRequestMetrics from the incoming data. In this example the metrics are logged to Console as JSON, but you can do anything with them, including logging specific data to NLog or keeping the data around for testing purposes.
```
public class TestTraceListener : TraceListener
{
    public override void TraceData(TraceEventCache eventCache, string source, TraceEventType eventType, int id, object data)
    {
        Log(eventCache, source, eventType, id, data);
    }
    public override void TraceData(TraceEventCache eventCache, string source, TraceEventType eventType, int id, params object[] data)
    {
        Log(eventCache, source, eventType, id, data);
    }
    public override void TraceEvent(TraceEventCache eventCache, string source, TraceEventType eventType, int id)
    {
        Log(eventCache, source, eventType, id);
    }
    public override void TraceEvent(TraceEventCache eventCache, string source, TraceEventType eventType, int id, string format, params object[] args)
    {
        Log(eventCache, source, eventType, id, string.Format(format, args));
    }
    public override void TraceEvent(TraceEventCache eventCache, string source, TraceEventType eventType, int id, string message)
    {
        Log(eventCache, source, eventType, id, message);
    }
    public override void TraceTransfer(TraceEventCache eventCache, string source, int id, string message, Guid relatedActivityId)
    {
        Log(eventCache, source, TraceEventType.Transfer, id, relatedActivityId);
    }
    public override void Write(string message)
    {
        TraceData(new TraceEventCache(), this.Name, TraceEventType.Information, 0, message);
    }
    public override void WriteLine(string message)
    {
        TraceData(new TraceEventCache(), this.Name, TraceEventType.Information, 0, message);
    }

    private void Log(TraceEventCache eventCache, string source, TraceEventType eventType, int eventId, params object[] data)
    {
        // first element is always LogMessage
        // second element may be an exception
        // https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/Amazon.Runtime/Internal/Util/_bcl/Logger.Diagnostic.cs#L48-L66
        var logMessage = data[0] as LogMessage;
        var exception = data.Length > 1 ? data[1] as Exception : null;
        // IRequestMetrics should be the first element in logMessage.Args
        // https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/Amazon.Runtime/Pipeline/PipelineHandler.cs#L174-L184
        var requestMetrics = (logMessage == null || logMessage.Args.Length == 0) ? null :
            logMessage.Args[0] as IRequestMetrics;

        if (requestMetrics != null)
        {
            Console.WriteLine("Got IRequestMetrics!");
            Console.WriteLine(requestMetrics.ToJSON());
        }
        else
        {
            Console.WriteLine("Got other data");
            Console.WriteLine(logMessage.ToString());
        }
        if (exception != null)
        {
            Console.WriteLine("Also got an exception");
            Console.WriteLine(exception);
        }
    }
}
```

And here you can see how it's possible to use the listener with the SDK:
```
// configure the SDK to use the listener we created
var listener = new TestTraceListener();
AWSConfigs.AddTraceListener("Amazon", listener);
AWSConfigs.LoggingConfig.LogMetrics = true;
AWSConfigs.LoggingConfig.LogResponses = ResponseLoggingOption.Always;
AWSConfigs.LoggingConfig.LogTo = LoggingOptions.SystemDiagnostics;

// make an AWS call to verify that the listener works
using(var ddb = new AmazonDynamoDBClient())
{
    ddb.ListTables();
}
```

And of course this approach will work if you use an in-memory log4net logger or similar TraceListener, then process the in-memory data in a similar approach to one I outlined.

This is a temporary workaround while we consider the full impact of implementing your suggestion. And of course we do accept pull requests, so if you're up for implementing the logging structure you described, we'd be more than happy to consider including it in the SDK.

**Comment by petemounce** [comment]
Beautiful! Thanks very much. Also for the blog post.

**Comment by PavelSafronov** [comment]
.NET SDK blog post related to this issue - http://blogs.aws.amazon.com/net/post/Tx2O1VF0PL83TGU

**Comment by toddb** [comment]
Sorry, to open up this issue again. I hope this is the right place. This solution is for the `net35` and `net45` builds and will not work the `netstandard1.3`. I an in aspnetcore 2.1 I have both looked at the source and decompiled versions.

As a summary:

- `AWSConfigs.AddTraceListener` no longer exists (see partial classes [_coreclr/AWSConfigs ](https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/_coreclr/AWSConfigs.coreclr.cs)AWSConfigs  and [AWSConfigs](https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/AWSConfigs.cs)
- That is congruent with the InternalSystemDiagnosticsLogger being compiled out (I can't work out why but the[ decompiled version confirms this source code]( https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/Amazon.Runtime/Internal/Util/Logger.cs#L48-51))

I currently have logging through to console.

```
AWSConfigs.LoggingConfig.LogTo = LoggingOptions.Console | LoggingOptions.SystemDiagnostics;
AWSConfigs.LoggingConfig.LogMetrics = true;
AWSConfigs.LoggingConfig.LogResponses = ResponseLoggingOption.Always;
AWSConfigs.LoggingConfig.LogMetricsFormat = LogMetricsFormatOption.JSON;

```

How can I intercept these because the `TraceListener` is no longer a solution yet the `LoggingOptions.SystemDiagnostics` still exists. I don't use log4net but rather NLog (and this isn't unreasonable IMHO).

Any thoughts on strategy either for me or what is happening with the sdk? Is this actually a compiliing bug? Thanks in advance
